driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -78,12 +78,12 @@
                           final ConnectionPoolSettings settings) {
         this.serverId = notNull("serverId", serverId);
         this.settings = notNull("settings", settings);
-        UsageTrackingInternalConnectionItemFactory connectionItemFactory
-        = new UsageTrackingInternalConnectionItemFactory(internalConnectionFactory);
+        UsageTrackingInternalConnectionItemFactory connectionItemFactory =
+                new UsageTrackingInternalConnectionItemFactory(internalConnectionFactory);
         pool = new ConcurrentPool<UsageTrackingInternalConnection>(settings.getMaxSize(), connectionItemFactory);
+        this.connectionPoolListener = getConnectionPoolListener(settings);
```

Comment:
There's a lurking race condition here: the `connectionPoolListener` field was assigned after the created of the maintenance timer thread, which uses it.  If the thread starts before assigning the field, the result is an NPE (which I observed several times while running the spec tests).  Moving the assignment up makes the race condition less likely, but there is still a race: the field is final, but the VM is not required to "publish" the field to all thread until the constructor completes.  So it's still possible that the maintenance timer thread will see a null value when accessing this field.

I think I'll try to fix this in a subsequent patch.

Reply:
ACK